src/datanode/src/heartbeat/task_tracker.rs

```diff
@@ -0,0 +1,277 @@
+pub(crate) enum RegisterResult<T: Send + Sync + Clone> {
+    // The watcher of the running task.
+    Busy(TaskWatcher<T>),
+    // The watcher of the newly registered task.
+    Running(TaskWatcher<T>),
+}
+
+impl<T: Send + Sync + Clone> RegisterResult<T> {
+    pub(crate) fn into_watcher(self) -> TaskWatcher<T> {
+        match self {
+            RegisterResult::Busy(inner) => inner,
+            RegisterResult::Running(inner) => inner,
+        }
+    }
+
+    /// Returns true if it's [RegisterResult::Busy].
+    pub(crate) fn is_busy(&self) -> bool {
+        matches!(self, RegisterResult::Busy(_))
+    }
+
+    #[cfg(test)]
+    /// Returns true if it's [RegisterResult::Running].
+    pub(crate) fn is_running(&self) -> bool {
+        matches!(self, RegisterResult::Running(_))
+    }
+}
+
+/// The result of waiting.
+pub(crate) enum WaitResult<T> {
+    Timeout,
+    Finish(Result<T>),
+}
+
+#[cfg(test)]
+impl<T> WaitResult<T> {
+    /// Returns true if it's [WaitResult::Timeout].
+    pub(crate) fn is_timeout(&self) -> bool {
+        matches!(self, WaitResult::Timeout)
+    }
+
+    /// Into the [WaitResult::Timeout] if it's.
+    pub(crate) fn into_finish(self) -> Option<Result<T>> {
+        match self {
+            WaitResult::Timeout => None,
+            WaitResult::Finish(result) => Some(result),
+        }
+    }
+}
+
+impl<T: Send + Sync + Clone + 'static> TaskTracker<T> {
+    /// Returns an empty [AsyncTaskTracker].
+    pub(crate) fn new() -> Self {
+        Self {
+            inner: Arc::new(RwLock::new(TaskTrackerInner::default())),
+        }
+    }
+
+    /// Waits for a [RegisterResult] and returns a [WaitResult].
+    pub(crate) async fn wait(
+        &self,
+        watcher: &mut TaskWatcher<T>,
+        timeout: Duration,
+    ) -> WaitResult<T> {
+        match tokio::time::timeout(timeout, wait(watcher)).await {
+            Ok(result) => WaitResult::Finish(result),
+            Err(_) => WaitResult::Timeout,
+        }
+    }
+
+    /// Tries to register a new async task, returns [RegisterResult::Busy] if previous task is running.
+    pub(crate) async fn try_register(
+        &self,
+        region_id: RegionId,
+        fut: BoxFuture<'static, Result<T>>,
+    ) -> RegisterResult<T> {
+        let mut inner = self.inner.write().await;
+        if let Some(task) = inner.state.get(&region_id) {
+            RegisterResult::Busy(task.watcher.clone())
+        } else {
+            let moved_inner = self.inner.clone();
+            let (tx, rx) = watch::channel(TaskState::<T>::Running);
+            common_runtime::spawn_bg(async move {
+                match fut.await {
+                    Ok(result) => {
+                        let _ = tx.send(TaskState::Done(result));
+                    }
+                    Err(err) => {
+                        let _ = tx.send(TaskState::Error(Arc::new(err)));
+                    }
+                };
+                moved_inner.write().await.state.remove(&region_id);
```

Comment:
If common runtime is single threaded, will this cause dead lock?

Reply:
It won't. It acquires the lock in the future.